### PR TITLE
当设计模式遇上 Kotlin

### PR DESCRIPTION
**Pull request by boileryao**
Gang of Four Patterns in Kotlin, 设计模式遇上 Kotlin
___________________
翻译完成

### CODE REVIEW

**Comment by windmxf** [comment]
校对认领 @sqrthree

**Comment by sqrthree** [comment]
@windmxf 好哒

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 1]
```diff
@@ -1,44 +1,44 @@
...
 
-If you don't like reading explanations, you can directly jump to my [github repository](https://github.com/lmller/gof-in-kotlin)
+如果不想看下面这坨文字的话，你可以直接去 [这个 GitHub 仓库](https://github.com/lmller/gof-in-kotlin) 看代码。
 
 ---
 
-As you might know, there are three types of (gof) patterns: **structural**, **creational** and **behavioral** patterns.
+众所周知，根据“四人帮”的定义设计模式可以分为三种: **结构型**、**创建型 **和 **行为型**。
 
-First up are the structural patterns. This is tough, because structural patterns are about - structure! How could I implement a structure with a *different* structure? I can't. An exception is the **Decorator**. While it's technically just about structure, its usage is more about behavior or responsibilities.
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [its usage is more about behavior or responsibilities]=>[但它的使用上，更多的和行为及职责有关。]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 2]
```diff
@@ -1,44 +1,44 @@
...
 ```
 class Text(val text: String) {
     fun draw() = print(text)
 }
 ```
 
-If you know the pattern, you also know that a set of classes has to be created in order to "decorate" (that is extending the behavior) the `Text` class.
+如果了解这个模式的话，你应该知道我们需要创建一些类来“修饰”（即，拓展行 为） `Text` 类。
 
-These extra classes can be avoided in Kotlin by using *extension functions*:
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [These extra classes]=>[这些额外的类]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 3]
```diff
@@ -128,13 +128,13 @@ println("Email1 goes to " + mail.recipient + " with subject " + mail.subject)
 println("Email2 goes to " + copy.recipient + " with subject " + copy.subject)
 ```
 
-#### Singleton
+#### 单例模式（Singleton）
 
-> Ensure a class only has one instance, and provide a global point of access to it
+> 确保一个类只有一个实例，并提供这个实例的全局访问点
 
-Although the **Singleton** is considered an anti-pattern these days, it has its usages (I won't discuss this topic here, just use it with caution).
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [just use it with caution]=>[但需要谨慎地使用它。]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 4]
```diff
@@ -197,13 +197,13 @@ execute {
 }
 ```
 
-As you can see, there is no need to have a class at all! One could argue, that this now resembles the **Strategy** pattern, which is probably correct. But then again, **Strategy** and **Template Method** are solving very similar problems (if not the same).
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [As you can see, there is no need to have a class at all!]=>[可以看到，这里不需要写任何一个类！]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 5]
```diff
@@ -229,33 +229,33 @@ println("${student.name} pays %.2f per month".format(student.pricePerMonth()))
 println("${regular.name} pays %.2f per month".format(regular.pricePerMonth()))
 ```
 
-#### Iterator
+#### 迭代器模式（Iterator）
 
-> Provide a way to access the elements of an aggregate object sequentially without exposing its underlying representation.
+> 提供了一种在不暴露其底层表示的情况下顺序访问聚合对象内部元素的方法
 
-Writing an **Iterator** is a rare task. Most of the time, it's easier and more convenient to wrap a `List` and implement the `Iterable` interface.
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [Writing an **Iterator** is a rare task]=>[很少需要自己写一个**迭代器**。]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 6]
```diff
@@ -229,33 +229,33 @@ println("${student.name} pays %.2f per month".format(student.pricePerMonth()))
...
+很难遇到手搓一个 **迭代器** 的情况。大多数情况，包装一个 `List` 并且实现 `Iterable`接口要更简单方便。
 
-In Kotlin, `iterator()` is an operator function. This means that when a class defines a function `operator fun iterator()`, it can be iterated using a `for` loop (no interface needed). That's particularly cool because it works also with extension functions. That's right - by using an extension function, I can make *every* object iterable. Consider this example:
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [That's particularly cool]=>[这种方式非常酷。]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 7]
```diff
@@ -229,33 +229,33 @@ println("${student.name} pays %.2f per month".format(student.pricePerMonth()))
...
-I can now iterate over a `Sentence`. This also works if I'm not the owner of the class.
+现在我们可以在 `Sentence` 上进行迭代操作了。如果没有这个类的控制权的话，迭代器仍然将正常工作。
 
-### More patterns
+### 更多的模式……
 
-I mentioned a few patterns, but those are not all *Gang of Four* patterns. As I said in the introduction, especially the structural patterns are hard or impossible to write in a different way than in Java. [Some other patterns can be found in the repository](https://github.com/lmller/gof-in-kotlin). I'm open for feedback and pull-requests ☺.
+这篇文章确实提到了相当几个设计模式，但这不是 **“四人帮”** 设计模式的全部。就像我在一开始提到的那样，尤其是结构型设计模式很难甚至根本不可能用和 Java 不同的方法来实现。 你可以在 [这个代码仓库](https://github.com/lmller/gof-in-kotlin) 找到更多的设计模式。欢迎来提交反馈和 PR。☺
 
-I hope this post gave you an idea on how Kotlin can result in different approaches to well known problems.
```
Comment (windmxf) [inline TODO/gang-of-four-patterns-in-kotlin.md]: [-I hope this post gave you an idea on how Kotlin can result in different approaches to well known problems.]=>[希望这篇文章能给你些启发，让你了解用Kotlin可以带来新的思路来解决那些广为人知的问题。]

**Comment by wilsonandusa** [comment]
xiao dui ren ling @sqrthree

**Comment by sqrthree** [comment]
@wilsonandusa 好哒

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 8]
```diff
@@ -1,44 +1,44 @@
...
+受到 Mario Fusco 的“从‘四人帮’到 lambda”（相关的[视频](https://www.youtube.com/watch?v=Rmer37g9AZM)、[博客](https://www.voxxed.com/blog/2016/04/gang-four-patterns-functional-light-part-1/)、[代码](https://github.com/mariofusco/from-gof-to-lambda)）的启发，我决定动手实现一些计算机科学领域最著名的设计模式，用 “Kotlin”！（“四人帮”指 Erich Gamma、Richard Helm、Ralph Johnson 和 John Vlissides，四人在所著的《Design Patterns: Elements of Reusable Object-Oriented Software 》一书中介绍了 23 种设计模式，该书被誉为设计模式的经典之作。——译注）
 
-The goal is not to simply *implement* the patterns, though. Since Kotlin supports object oriented programming and is interoperable with Java, I could just copy-auto-convert every java class in Mario's repository (doesn't matter whether it's the "traditional" or the "lambda" examples) and **it would still work!**
+当然，我的目标不是简单的 **实现** 这些模式。因为 Kotlin 支持面向对象编程并且和 Java 是可互操作的，我可以从 Mario 的仓库直接复制粘贴每一个 Java 文件（先不管是“传统”的还是“lambada 风格”的），**它们将仍然可以正常工作**！
```
Comment (wilsonandusa) [inline TODO/gang-of-four-patterns-in-kotlin.md]: typo: [lambada] => [lambda]

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 9]
```diff
@@ -1,44 +1,44 @@
...
 
-It's also important to note that these patterns where discovered to overcome the shortcomings of how imperative programming languages were designed in the 90s (C++ in particular). Many modern languages provide features to overcome these issues without writing extra code or falling back to patterns.
+需要特别说明一下，这些模式的发明是为了弥补起源于上世纪九十年代的一些命令式编程语言（尤其是 C++）的不足。很多现代编程语言提供了解决这些不足的特性，我们完全不需要再写多余的代码或者做刻意模仿设计模式这种事了。
 
-That's why —just like Mario in the `g ∘ f` repository— I will try to find a simpler, easier or more idiomatic way to solve the same problem each pattern is solving.
+这就是为什么我像 Mario 那样，去寻找一种更简单方便、更惯用的方式来解决这些模式所要解决的问题。
```
Comment (wilsonandusa) [inline TODO/gang-of-four-patterns-in-kotlin.md]: g of f 地址(https://github.com/mariofusco/from-gof-to-lambda) 可以考虑加上

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 10]
```diff
@@ -1,44 +1,44 @@
 > * 原文地址：[Gang of Four Patterns in Kotlin](https://dev.to/lovis/gang-of-four-patterns-in-kotlin)
 > * 原文作者：[Lovis](https://dev.to/lovis)
 > * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
-> * 译者：
+> * 译者：[Boiler Yao](https://github.com/boileryao)
 > * 校对者：
```
Comment (wilsonandusa) [inline TODO/gang-of-four-patterns-in-kotlin.md]: 校对者：[wilsonandusa](https://github.com/wilsonandusa)

**Comment by BoilerYao** [comment]
校对完成 @sqrthree

**Comment by sqrthree** [comment]
@BoilerYao @windmxf 的 review 还有许多没有折叠，是不需要调整吗？还是需要讨论下？麻烦 @windmxf 确认下哈。

File: TODO/gang-of-four-patterns-in-kotlin.md [hunk 11]
```diff
@@ -1,44 +1,44 @@
...
-That's why —just like Mario in the `g ∘ f` repository— I will try to find a simpler, easier or more idiomatic way to solve the same problem each pattern is solving.
+这就是为什么我像 Mario （相关仓库地址：[gof](https://github.com/mariofusco/from-gof-to-lambda)）那样，去寻找一种更简单方便、更惯用的方式来解决这些模式所要解决的问题。
 
-If you don't like reading explanations, you can directly jump to my [github repository](https://github.com/lmller/gof-in-kotlin)
+如果不想看下面这坨说明文字的话，你可以直接去 [这个 GitHub 仓库](https://github.com/lmller/gof-in-kotlin) 看代码。
 
 ---
 
-As you might know, there are three types of (gof) patterns: **structural**, **creational** and **behavioral** patterns.
+众所周知，根据“四人帮”的定义设计模式可以分为三种: **结构型**、**创建型 **和 **行为型**。
```
Comment (sqrthree) [inline TODO/gang-of-four-patterns-in-kotlin.md]: image

**Comment by BoilerYao** [comment]
好啦 @sqrthree 
掘金地址：https://juejin.im/post/594b2ac00ce4630057425bd5

**Pull request merged**